flink-runtime/src/test/java/org/apache/flink/runtime/asyncprocessing/AsyncExecutionControllerTest.java

```diff
@@ -0,0 +1,305 @@
+        assertThat(output.get()).isEqualTo(1);
+        assertThat(recordContext4.getReferenceCount()).isEqualTo(0);
+    }
+
+    /**
+     * An AsyncExecutionController for testing purpose, which integrates with basic buffer
+     * mechanism.
+     */
+    static class TestAsyncExecutionController<R, K> extends AsyncExecutionController<R, K> {
+
+        LinkedList<StateRequest<K, ?, ?>> activeBuffer;
+
+        LinkedList<StateRequest<K, ?, ?>> blockingBuffer;
+
+        public TestAsyncExecutionController(
+                MailboxExecutor mailboxExecutor, StateExecutor stateExecutor) {
+            super(mailboxExecutor, stateExecutor);
+            activeBuffer = new LinkedList<>();
+            blockingBuffer = new LinkedList<>();
+        }
+
+        @Override
+        <IN, OUT> void insertActiveBuffer(StateRequest<K, IN, OUT> request) {
+            activeBuffer.push(request);
+        }
+
+        <IN, OUT> void insertBlockingBuffer(StateRequest<K, IN, OUT> request) {
+            blockingBuffer.push(request);
+        }
+
+        void triggerIfNeeded(boolean force) {
+            if (!force) {
+                // Disable normal trigger, to perform single-step debugging and check.
+                return;
+            }
+            LinkedList<StateRequest<?, ?, ?>> toRun = new LinkedList<>(activeBuffer);
+            activeBuffer.clear();
+            stateExecutor.executeBatchRequests(toRun);
+        }
+
+        void migrateBlockingToActive() {
+            Iterator<StateRequest<K, ?, ?>> blockingIter = blockingBuffer.iterator();
+            while (blockingIter.hasNext()) {
+                StateRequest<K, ?, ?> request = blockingIter.next();
+                if (request.getRecordContext().tryOccupyKey()) {
+                    insertActiveBuffer(request);
+                    blockingIter.remove();
+                }
+            }
+        }
+    }
+
+    /** Simulate the underlying state that is actually used to execute the request. */
+    static class TestUnderlyingState {
+
+        private HashMap<String, Integer> hashMap;
+
+        public TestUnderlyingState() {
+            this.hashMap = new HashMap<>();
+        }
+
+        public Integer get(String key) {
+            return hashMap.get(key);
+        }
+
+        public void update(String key, Integer val) {
+            hashMap.put(key, val);
+        }
+    }
+
+    static class TestValueState implements ValueState<Integer> {
+
+        private AsyncExecutionController<String, String> asyncExecutionController;
+
+        private TestUnderlyingState underlyingState;
+
+        public TestValueState(
+                AsyncExecutionController<String, String> aec, TestUnderlyingState underlyingState) {
+            this.asyncExecutionController = aec;
+            this.underlyingState = underlyingState;
+        }
+
+        @Override
+        public StateFuture<Void> asyncClear() {
+            StateRequest<String, Void, Void> request =
```

Comment:
Generate `StateRequest` in State layer looks good to me.
But how could we handle `SYNC` type which is invisible for `State` ?

Reply:
`SYNC` is special case that will be only generated by the framework (only before the `processElement`). The direct implementation of State API should not create this kind of request.